### ISSUE TITLE
Questions, questions!

### ISSUE THREAD

**Issue by bartavelle** [folq/haskell-to-elm#9]
Hello,

  I am the current maintainer of elm-bridge. I was looking for a Elm pretty printer to replace the terrible string-concat code that is currently emitting code, only to realize there is another library that tries to support aeson's options! Instead of duplicating work, I'd rather deprecate elm-bridge and point users towards haskell-to-elm. However I have a couple questions:

 * are you going to support this library, or are you just experimenting?
 * what is missing exactly wrt. aeson's Options support? elm-bridge should support it all (it is currently missing the *JsonKey stuff).
 * would you accept contributions for missing features?

**Comment by ollef** [comment]
Hey!

That sounds fine to me. `elm-bridge` and a few other libraries were a great inspiration for `haskell-to-elm`, but we felt it would be easier to start from scratch given the design and tight integration with a Servant client library generator (`servant-to-elm`) that we were after.

1. We're using this quite heavily in production at Folq, which is a Norwegian startup company, so we're going to support it at least for the foreseeable future.
1. I created https://github.com/folq/haskell-to-elm/issues/10 for tracking the support of all Aeson options. In summary, the current situation is that the code takes into account all options except for different `sumEncoding`s, but there aren't yet tests for anything that's not `defaultOptions`.
1. Contributions would be very welcome.

**Comment by bartavelle** [comment]
On the testing side, I think the best thing in `elm-bridge` is [this test](https://github.com/agrafix/elm-bridge/blob/28400342b143f7899e82fa22a435461116d38760/test/EndToEnd.hs). It generates encodings for many option combinations using `Arbitrary`, and generates an [Elm test file](https://github.com/agrafix/elm-bridge/blob/28400342b143f7899e82fa22a435461116d38760/test/current-end-to-end/src/MyTests.elm). That way you can property test encoding/decoding. I can try my hand at porting it to haskell-to-elm. What do you think?

**Comment by ollef** [comment]
That'd be a _very_ welcome contribution! Generating the test cases is a much better idea than doing it mostly by hand like in haskell-to-elm-test. Perhaps it could even be nearly exhaustively tested (with some limitations like testing only with a finite number of label modifier functions).

**Comment by bartavelle** [comment]
Right now I hand write a finite number of types, and `Options`, but the "contents" get generated by Arbitrary. I suppose one could generate arbitrary types in the first place, that would be a much better solution!

**Comment by ollef** [comment]
That also sounds like a good idea. I was thinking that the `Options` could be (almost) exhaustively tested, i.e. that we don't need to use `Arbitrary` to generate different options, but can produce a list of all of them, with the limitation that we only test a few different label modifier functions.

**Comment by bartavelle** [comment]
So I tried porting the test suite, and I found that there is still a bit more work left before I can deprecate `elm-bridge`, such as support for more basic types (I saw the PR).

However, the entire point of `elm-bridge` is to generate compatible `aeson`/Elm functions by construction. For this to work, I think TH is a great choice for *users*, as instead of manually writing 6 instances for `haskell-to-elm` (for parameterized types) and 2 instances to `aeson`, making sure they are using the same set of options every time, they would only have to write a single line.

The generics code is much nicer to write though, and some people have a religious aversion to TH, so I can see how that makes senses to use Generics.

There is also something I did not understand:

```haskell
data Foo
  = Foo
  { _fooX1 :: Int
  , _fooX2 :: Int
  }
```

When deriving using `fieldLabelModifier = drop 4 `, the JSON fields will not be `X1`, but `_fooX1`. Is that expected?

**Comment by ollef** [comment]
> support for more basic types (I saw the PR).

What basic types are missing? That is usually quite easy to fix.

> generate compatible aeson/Elm functions by construction.

Yeah, this is a little problematic when using haskell-to-elm because it's possible to choose incompatible Aeson options in the different instances.

We've been using DerivingVia to derive all the instances using the same options. As a first fix to this problem, we can perhaps recommend using that and show how to do it in the documentation (https://github.com/folq/haskell-to-elm/issues/6). As a better fix, perhaps providing types to derive via could work?

> When deriving using fieldLabelModifier = drop 4 , the JSON fields will not be X1, but _fooX1. Is that expected?

There are two `fieldLabelModifier`s. The one in `Aeson.Options` controls the JSON field names. The one from the `Options` type from this library controls the field names in the Elm code, which can be different from both the Haskell and the JSON field names. If you're seeing some other behaviour it might be a bug. :)

**Comment by bartavelle** [comment]
> What basic types are missing? That is usually quite easy to fix.

My first test had a `Map` in it :)

**Comment by ollef** [comment]
Yeah, I don't know of a satisfactory solution for Map. :/

**Comment by bartavelle** [comment]
I am not really familiar with how "deriving via" works, and it would indeed be quite nice to have an example here :)

**Comment by ollef** [comment]
I've added some docs for it to the readme now. :)

**Comment by bartavelle** [comment]
How can you set specific Aeson options that way? That is the thing I do not understand, how Aeson interacts with haskell-to-elm. Sorry for being so dense :)

**Comment by ollef** [comment]
The options are set in the instances of the type that you derive via (`ElmType` in [the example](https://github.com/folq/haskell-to-elm/blob/master/examples/DerivingVia.hs)), which you can often (at least in my experience) use to derive instances for most of your project.

A more flexible way to specify options with deriving via would be to do something like https://hackage.haskell.org/package/deriving-aeson, which gives you a type to derive via which allows specifying different Aeson options on the type level.

**Comment by bartavelle** [comment]
Alright, I think I get the gist of it, but I am having trouble making it work with parameterized types, even with standalone deriving declarations. Still working on it though!

**Comment by bartavelle** [comment]
Relying on `UndecidableInstances` is pretty bad in the presence of parameterized types.
I am still looking into this, but I think I will write the instances manually from now on.

**Comment by ollef** [comment]
Why is it bad?

**Comment by bartavelle** [comment]
Because it prevents the code from compiling, whereas it should compile with hand defined instances and not the UndecidableInstances switch. I will have to work a bit more on this!

**Comment by ollef** [comment]
Hmm, do you need it to use the library? [The example](https://github.com/folq/haskell-to-elm/blob/master/examples/Parameterised.hs) doesn't.

Either way, IMO `UndecidableInstances` is quite a benign extension to use. It sounds scarier than it is. :)

**Comment by bartavelle** [comment]
Indeed, the example doesn't use it, because the instances are not derived.
